### PR TITLE
lxd: Add support for starting instances on creation (from Incus)

### LINKED ISSUE #13322: Add create and start endpoint
# Required information

 * Distribution: 5.21 LTS
 * Distribution version: snap

# Issue description

Please add the option to start an instance to the `POST 1.0/instances` API endpoint. This could be an additional query parameter like `?start=true` or expressed in the payload.

Motivation is LXD-UI. It has an option to "create and start" an instance, which under the hood posts to `1.0/instances`. Once the instance is created, the UI changes the state of the instance via `PUT 1.0/instances/:instance/state`. This is the same as the CLI is doing it. But in the UI this is problematic, because a user might refresh the page in between -- i.e. when the creation is taking a long time because the image needs to be fetched first. In that case the ui is missing to send the 2nd request, because the callback to do so is lost due to the refresh operation.

### PR DESCRIPTION
**Pull request by boltmark**
This PR adds support for starting instances on creation.

Includes cherry-picks from https://github.com/lxc/incus/pull/410.

Closes #13322

### CODE REVIEW

**Comment by ru-fu** [comment]
This will also need a doc update to the API tab for creating an instance: https://documentation.ubuntu.com/lxd/en/latest/howto/instances_create/
The other tabs say that you have the choice between init & launch / create & create and start, while the API says you have to start the instance separately.

You can find a short intro to how the docs work [here](https://documentation.ubuntu.com/lxd/en/latest/contributing/#contribute-to-the-documentation) - get in touch if anything is unclear. :)

**Comment by tomponline** [comment]
Hi @boltmark looks like not all of the commits have been cherry-picked correctly as the CLA test is failing.

Also please can you sign your commits, see https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification

Thanks!

**Comment by boltmark** [comment]
Thanks for the comments.
@tomponline -- CLA test is now passing.
@ru-fu -- I updated the documentation to include information + an example on starting containers upon creation.

File: doc/howto/instances_create.md [hunk 1]
```diff
@@ -72,6 +72,19 @@ To start an instance, send a PUT request to change the instance state:
...
+      "source": {
+        "alias": "<image_alias>",
+        "protocol": "simplestreams",
+        "server": "<server_URL>",
+        "type": "image"
+      }
+      "start": true
+    }'
+
 See {ref}`instances-manage-start` for more information.
```
Comment (ru-fu) [inline doc/howto/instances_create.md:88]: This link is about the command for starting the instance, so it should go above your addition.

File: doc/api-extensions.md [hunk 2]
```diff
@@ -2424,3 +2424,10 @@ The OVN driver will allocate IP addresses from the subnets specified in the upli
 
 Adds the ability to explicitly specify a trust token when creating a certificate
 and joining an existing cluster.
+
+## `instance_create_start`
+
+Adds a new `start` field to the `POST /1.0/instances` API which when set
```
Comment (ru-fu) [inline doc/api-extensions.md:2430]: Sorry, nitpicking here. ;)

```suggestion
Adds a `start` field to the `POST /1.0/instances` API which when set
```

You wouldn't add an old field, would you? :wink:

**Review by ru-fu — COMMENTED** [review]
Thanks, this looks good! Two small comments (but none are blocking).

File: lxd/instances_post.go [hunk 3]
```diff
@@ -1357,3 +1372,17 @@ func clusterCopyContainerInternal(s *state.State, r *http.Request, source instan
 	// Run the migration
 	return createFromMigration(s, nil, projectName, profiles, req)
 }
+
+func instanceCreateFinish(s *state.State, req *api.InstancesPost, args db.InstanceArgs) error {
```
Comment (tomponline) [inline lxd/instances_post.go:1376]: Missing doc block would be useful here.

File: lxd/instances_post.go [hunk 4]
```diff
@@ -377,7 +388,8 @@ func createFromMigration(s *state.State, r *http.Request, projectName string, pr
 
 		instOp.Done(nil) // Complete operation that was created earlier, to release lock.
 		runRevert.Success()
-		return nil
+
+		return instanceCreateFinish(s, req, args)
```
Comment (tomponline) [inline lxd/instances_post.go:392]: So I think this has the potential to cause problems when doing a stateful (live) migration, as the instance is naturally restored on the target in a running (started) state, so if you request it be started it will fail the migration because its already running.

Please can you running LXD in 2 VMs, add a remote so they can talk to it other using `lxc remote add` and then try live migrating a nested VM (a VM running inside one of the LXD VMs) to the other run, but with this new `Start` field set to `true` and see if it breaks.

If you need help setting this up let me know and we can arrange a 1:1 video call to go through it.

File: doc/howto/instances_create.md (hunk 1 above)
Comment (boltmark) [reply doc/howto/instances_create.md]: Ah, thanks. Moved the link back to original location.

File: doc/api-extensions.md (hunk 2 above)
Comment (boltmark) [reply doc/api-extensions.md]: I agree. Fixed now :)

File: lxd/instances_post.go (hunk 3 above)
Comment (boltmark) [reply lxd/instances_post.go:1509]: Thanks, added comment.

File: lxd/instances_post.go (hunk 4 above)
Comment (boltmark) [reply lxd/instances_post.go]: As you expected, this caused failure for stateful migration. I added a check in `instanceCreateFinish` to ensure we do not attempt to start a container that is already running.

Comment (tomponline) [reply lxd/instances_post.go]: thanks! Let me know when ready for review.

**Comment by boltmark** [comment]
@tomponline Ready for review. Looks like the documentation test is failing due to criu.org being down, but hopefully this is temporary.

File: lxc/init.go [hunk 5]
```diff
@@ -11,7 +11,7 @@ import (
 	"github.com/spf13/cobra"
 	"gopkg.in/yaml.v2"
 
-	"github.com/canonical/lxd/client"
+	lxd "github.com/canonical/lxd/client"
```
Comment (tomponline) [inline lxc/init.go:14]: @boltmark please can you not auto complete the `client` package as `lxd` here. Thanks

File: doc/api-extensions.md [hunk 6]
```diff
@@ -2437,7 +2437,7 @@ Adds the ability to convert images from different formats (e.g. VMDK or QCow2) i
 
 ## `instance_create_start`
 
-Adds a new `start` field to the `POST /1.0/instances` API which when set
+Adds a `start` field to the `POST /1.0/instances` API which when set
```
Comment (tomponline) [inline doc/api-extensions.md:2440]: Please can you reword this in https://github.com/canonical/lxd/pull/13695/commits/00b07d1b6468dbc15e34f8eeb0a12bc9e5e33462 and keep the doc change below as a separate commit

File: lxd/instances_post.go [hunk 7]
```diff
@@ -1504,16 +1504,22 @@ func clusterCopyContainerInternal(s *state.State, r *http.Request, source instan
...
 	}
 
-	// Start the instance.
+	// Load instance and start if not already running.
 	inst, err := instance.LoadByProjectAndName(s, args.Project, args.Name)
 	if err != nil {
 		return fmt.Errorf("Failed to load the instance: %w", err)
 	}
 
+	if inst.IsRunning() {
```
Comment (tomponline) [inline lxd/instances_post.go:1520]: This call can be relatively expensive, can we avoid needing this if we just dont try and start it when doing a live/stateful migration?

**Review by tomponline — CHANGES_REQUESTED** [review]
Couple of nits and question about the migration fix

File: lxd/instances_post.go (hunk 7 above)
Comment (boltmark) [reply lxd/instances_post.go]: I've updated `createFromMigration` to maintain its current behavior (i.e. not invoking the new `instanceCreateFinish` function), which means we no longer need the `inst.IsRunning` check.

File: lxc/init.go (hunk 5 above)
Comment (boltmark) [reply lxc/init.go]: Oops, fixed

File: doc/api-extensions.md (hunk 6 above)
Comment (boltmark) [reply doc/api-extensions.md:2440]: Done, thanks

**Comment by boltmark** [comment]
> Couple of nits and question about the migration fix

Thanks again @tomponline! I've resolved the nits and provided an alternative fix for the migration issue. Let me know what you think.

**Review by tomponline — APPROVED** [review]
Thanks!
